Storage-models.md

```diff
@@ -0,0 +1,393 @@
+
+  Note that `datastore_services` will not include everything from ndb. For example, only some properties are included. This is intentional! The properties in `datastore_services` are the only ones we support, so if you can't achieve what you want through `datastore_services`, you should not fall back to ndb. Instead, you should devise an alternate approach that only requires the functionality available through `datastore_services`.
+
+* `import_transaction_services()` returns a module that contains the decorator `run_in_transaction_wrapper`. You can use it to perform a datastore operation as a _transaction_, which means that it is guaranteed to either complete fully or not happen at all. In other words, if a transaction fails, the datastore will look like the transaction was never started.
+
+  Transactions are useful for keeping the datastore in a self-consistent state. For example, suppose you have both user models and user profile models, but you never want to have a profile without a corresponding user or a user with no associated profile. To prevent this, you can use transactions like this:
+
+  ```python
+  @transaction_services.run_in_transaction_wrapper
+  def delete_user(user_key, profile_key):
+      user_key.delete()
+      profile_key.delete()
+  ```
+
+  Now if `profile_key.delete()` fails, the transaction will ensure that the user doesn't get deleted either.
+
+Note that in some places the datastore interface is not well-defined. For example, `Registry` doesn't specify which ndb properties should be available. In practice, we assume that the interface is what our Google Appengine (GAE) implementation provides.
+
+#### The GAE datastore interface implementation
+
+The GAE implementation provides modules that the `Registry` functions return.
+
+* `import_models(model_names)`: Loading each module is handled by dedicated code in `_Gae.import_models()`. This means that if you add a new model module, you need to update `import_models()` to support it.
+
+  By convention, models are defined in files whose paths have the form `core/storage/<model module name>/gae_models.py`. For example, above we saw `base_models` being loaded using `import_models`. This module is defined at `core/storage/base_model/gae_models.py`, so `base_models` in the example above refers to `core.storage.base_model.gae_models`.
+
+* `import_datastore_services()`: The returned module is [`core.platform.datastore.cloud_datastore_services`](https://github.com/oppia/oppia/blob/develop/core/platform/datastore/cloud_datastore_services.py). You can refer to this module file to check which ndb properties we support.
+
+* `import_transaction_services()` returns the module [`core/platform/transactions/cloud_transaction_services.py`](https://github.com/oppia/oppia/blob/develop/core/platform/transactions/cloud_transaction_services.py).
```

Comment:
I've removed the code references, but I kept the rest because I think it's useful to explain the naming conventions and how to find which ndb properties we support